From 6e2c1e405b1e6532e5dc5ea804362d05989d27d6 Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:39:06 -0500
Subject: [drm] nvc0-/fb/ram: fix use of non-existant ram if partitions aren't uniform

Message-id: <1418852380-13061-56-git-send-email-rclark@redhat.com>
Patchwork-id: 102186
O-Subject: [RHEL7 drm 55/89] drm/nvc0-/fb/ram: fix use of non-existant ram if partitions aren't uniform
Bugzilla: 1173317
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1173317
Upstream: 3.17

commit 6c9825efa54b23ad0500a3744a25d8b4bdb2df1a

Author:     Ben Skeggs <bskeggs@redhat.com>
AuthorDate: Tue Aug 12 09:17:41 2014 +1000
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Wed Dec 10 17:45:29 2014 -0500

    drm/nvc0-/fb/ram: fix use of non-existant ram if partitions aren't uniform

    commit 3d9e3921f4d77bcaeea913c48b894d1208f0cb06 upstream

    Likely a large part of the GK106 woes..

    Signed-off-by: Ben Skeggs <bskeggs@redhat.com>

Signed-off-by: Rob Clark <rclark@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/nouveau/core/subdev/fb/ramnvc0.c b/drivers/gpu/drm/nouveau/core/subdev/fb/ramnvc0.c
index 5a6a502..cd99205 100644
--- a/drivers/gpu/drm/nouveau/core/subdev/fb/ramnvc0.c
+++ b/drivers/gpu/drm/nouveau/core/subdev/fb/ramnvc0.c
@@ -554,13 +554,13 @@ nvc0_ram_create_(struct nouveau_object *parent, struct nouveau_object *engine,
  } else {
   /* otherwise, address lowest common amount from 0GiB */
   ret = nouveau_mm_init(&pfb->vram, rsvd_head,
-          (bsize << 8) * parts, 1);
+          (bsize << 8) * parts - rsvd_head, 1);
   if (ret)
    return ret;
 
   /* and the rest starting from (8GiB + common_size) */
   offset = (0x0200000000ULL >> 12) + (bsize << 8);
-  length = (ram->size >> 12) - (bsize << 8) - rsvd_tail;
+  length = (ram->size >> 12) - ((bsize * parts) << 8) - rsvd_tail;
 
   ret = nouveau_mm_init(&pfb->vram, offset, length, 0);
   if (ret)
-- 
1.7.1